test/test_cuda.py

```diff
@@ -738,6 +738,50 @@ def test_caching_pinned_memory_multi_gpu(self):
         self.assertEqual(gpu_tensor1[0], 1)
         self.assertEqual(gpu_tensor0[0], 2)
 
+    def test_btrifact(self):
+        a = torch.Tensor((((1.3722, -0.9020),
+                           (1.8849, 1.9169)),
+                          ((0.7187, -1.1695),
+                           (-0.0139, 1.3572)),
+                          ((-1.6181, 0.7148),
+                           (1.3728, 0.1319)))).cuda()
+        LU_data, pivots, info = a.btrifact()
+        self.assertEqual(info.abs().sum(), 0)
+        I_U = torch.triu(torch.ones(2, 2)).unsqueeze(0).expand(3, 2, 2).type_as(a).byte()
+        I_L = 1 - I_U
+        a_L = torch.zeros(a.size()).type_as(a)
+        a_U = a_L.clone()
+        a_L[torch.eye(2).unsqueeze(0).expand(3, 2, 2).type_as(a).byte()] = 1.0
+        a_L[I_L] = LU_data[I_L]
+        a_U[I_U] = LU_data[I_U]
+
+        P = torch.eye(2).unsqueeze(0).expand(3, 2, 2).type_as(a)
+        for i in range(3):
+            for j in range(2):
+                k = pivots[i, j] - 1
+                t = P[i, j, :].clone()
+                P[i, j, :] = P[i, k, :]
+                P[i, k, :] = t
+
+        a_ = torch.bmm(P, torch.bmm(a_L, a_U))
+        self.assertEqual(a_, a)
+
+    def test_btrisolve(self):
+        a = torch.Tensor((((1.3722, -0.9020),
+                           (1.8849, 1.9169)),
+                          ((0.7187, -1.1695),
+                           (-0.0139, 1.3572)),
+                          ((-1.6181, 0.7148),
+                           (1.3728, 0.1319)))).cuda()
+        b = torch.Tensor(((4.02, 6.19),
+                          (-1.56, 4.00),
+                          (9.81, -4.09))).cuda()
+        LU_data, pivots, info = a.btrifact()
+        self.assertEqual(info.abs().sum(), 0)
```

Comment:
Yes, I think `btrifact` is slightly different than other Torch functions. The factorization can fail for some (but not necessarily all) of the examples in the minibatch. In our use of `btrifact` in `qpth`, we hit instability in some of the minibatch examples that causes them to not be LU-factorizable and returning `info` here provides a nice way of disabling them without failing the factorization for the entire minibatch.

I do agree that it's non-standard in Torch to return an `info` like this and require users to manually check the values. What do you think about making it an optional parameter? If it's provided, it can be written to as it is now, and otherwise `btrifact` can throw an error. If you prefer this, is an optional parameter easy to handle in `cwrap` or should I privatize the C call and add a quick Python wrapper for this?